contrib/generic_proxy/filters/network/source/proxy.cc

```diff
@@ -16,19 +16,59 @@ namespace Extensions {
 namespace NetworkFilters {
 namespace GenericProxy {
 
+namespace {
+
+Tracing::Decision tracingDecision(const Tracing::ConnectionManagerTracingConfig& tracing_config,
+                                  Runtime::Loader& runtime) {
+  bool traced = runtime.snapshot().featureEnabled("tracing.random_sampling",
+                                                  tracing_config.getRandomSampling());
+
+  if (traced) {
+    return {Tracing::Reason::Sampling, true};
+  }
+  return {Tracing::Reason::NotTraceable, false};
+}
+
+} // namespace
+
 ActiveStream::ActiveStream(Filter& parent, RequestPtr request)
-    : parent_(parent), downstream_request_stream_(std::move(request)) {}
+    : parent_(parent), downstream_request_stream_(std::move(request)),
+      stream_info_(parent_.time_source_,
+                   parent_.callbacks_->connection().connectionInfoProviderSharedPtr()) {
 
-ActiveStream::~ActiveStream() {
-  for (auto& filter : decoder_filters_) {
-    filter->filter_->onDestroy();
+  connection_manager_tracing_config_ = parent_.config_->tracingConfig();
+
+  auto tracer = parent_.config_->tracingProvider();
+
+  if (!connection_manager_tracing_config_.has_value() || !tracer.has_value()) {
+    return;
   }
-  for (auto& filter : encoder_filters_) {
-    if (filter->isDualFilter()) {
-      continue;
-    }
-    filter->filter_->onDestroy();
+
+  auto decision = tracingDecision(connection_manager_tracing_config_.value(), parent_.runtime_);
+  if (decision.traced) {
+    stream_info_.setTraceReason(decision.reason);
   }
+  active_span_ = tracer->startSpan(*this, *downstream_request_stream_, stream_info_, decision);
```

Comment:
Does `NullSpan` can be used when there is no config. Then we needn't check everywhere whether the active span is null or not.

Reply:
In my completely personal opinion, the null `active_span_` means no valid tracing config and do nothing for tracing. We cannot avoid all these checking. And there are only two positions at where `active_span_` will be checked. So I personally think it's ok. 🤔

Reply:
Actually you already use the nullspan at https://github.com/envoyproxy/envoy/pull/24790/files/824b4dae787d253e9eaab10a0dfec0d61dda30c0#diff-eb91cb182a2739c76dc1f18aa5ebfbb5139160a49e2490c2f46cf3de31010a80R237, but yes, it only save two checks, so I'm fine with it :)